nixos/modules/services/x11/redshift.nix

```diff
@@ -93,52 +34,59 @@ in {
-      type = types.listOf types.str;
-      default = [];
-      example = [ "-v" "-m randr" ];
-      description = ''
-        Additional command-line arguments to pass to
-        <command>redshift</command>.
-      '';
-    };
   };
 
   config = mkIf cfg.enable {
-    assertions = [ 
-      {
-        assertion = 
-          if cfg.provider == "manual"
-          then (cfg.latitude != null && cfg.longitude != null) 
-          else (cfg.latitude == null && cfg.longitude == null);
-        message = "Latitude and longitude must be provided together, and with provider set to null.";
-      }
-    ];
+    assertions =
+      with lib; let
+        allThe = x: !(any (map isNull x));
+        no = x: all (map isNull x);
+        oneOf = xs: (count id xs) == 1;
+        together = x: {
+          assertion = (allThe x.items) || (no x.items);
+          message = x.names + " must be provided together.";
+        };
+        coordinates = [ cfg.config.manual.lat cfg.config.manual.lon ];
+        times = [ cfg.config.redshift.dawn-time cfg.config.redshift.dusk-time ];
+        mechanisms = [ (allThe coordinates) (allThe times) ];
+        manualLocationProvider = cfg.config.redshift.location-provider == "manual";
+      in
+        (map together [ 
+          {
+            names = "Latitude and longitude";
+            items = coordinates;
+          }
+          {
+            names = "Time of dawn and time of dusk";
+            items = times;
+          }
+        ])
+        ++
+        [
+          {
+            assertion = oneOf mechanisms;
+            message = "The transitions must be set via exactly one mechanism.";
+          }
+          {
+            assertion = if manualLocationProvider then (allThe coordinates) else (no coordinates);
+            message = "Setting coordinates requires setting the location provider to manual and vice versa.";
+          }
+        ];
 
     # needed so that .desktop files are installed, which geoclue cares about
     environment.systemPackages = [ cfg.package ];
 
-    services.geoclue2.enable = mkIf (cfg.provider == "geoclue2") true;
+    services.geoclue2.enable = mkIf (cfg.config.redshift.location-provider == "geoclue2") true;
 
     systemd.user.services.redshift = 
-    let
-      providerString = 
-        if cfg.provider == "manual"
-        then "${cfg.latitude}:${cfg.longitude}"
-        else cfg.provider;
-    in
     {
       description = "Redshift colour temperature adjuster";
       wantedBy = [ "graphical-session.target" ];
       partOf = [ "graphical-session.target" ];
       serviceConfig = {
         ExecStart = ''
           ${cfg.package}/bin/redshift \
-            -l ${providerString} \
-            -t ${toString cfg.temperature.day}:${toString cfg.temperature.night} \
-            -b ${toString cfg.brightness.day}:${toString cfg.brightness.night} \
-            ${lib.strings.concatStringsSep " " cfg.extraOptions}
+            -c ${lib.toINI cfg.config}
```

Comment:
`lib` doesn't have a `toINI`, did you even evaluate this? It's in `lib.generators.toINI`